File: backend/app/api/schemas/project.py
"""Public HTTP response schema for projects."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict

from app.api.schemas.technology import TechnologyResponse


class ProjectResponse(BaseModel):
    """Serialized project and its technologies returned by the API."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    profile_id: UUID
    slug: str
    title: str
    description: str
    content: str
    github_url: str | None
    demo_url: str | None
    image_url: str | None
    is_featured: bool
    sort_order: int
    created_at: datetime
    updated_at: datetime
    technologies: list[TechnologyResponse]

File: frontend/src/schemas/project.ts
import { z } from "zod";

import { technologySchema } from "@/schemas/technology";

export const projectSchema = z.object({
  id: z.uuid(),
  profile_id: z.uuid(),
  slug: z.string(),
  title: z.string(),
  description: z.string(),
  content: z.string(),
  github_url: z.url().nullable(),
  demo_url: z.url().nullable(),
  image_url: z.url().nullable(),
  is_featured: z.boolean(),
  sort_order: z.number().int(),
  created_at: z.iso.datetime(),
  updated_at: z.iso.datetime(),
  technologies: z.array(technologySchema),
});

export const projectsSchema = z.array(projectSchema);
